[NanoAOD] Don't handle column types redundantly anymore

guitargeek opened the pull request:
#### PR description:

This is just a re-opening of https://github.com/cms-nanoAOD/cmssw/pull/401 from September 2019, but this time to upstream CMSSW instead of cmssw-nanoAOD. Back in the day I was interested in storing columns of so far not supported basic types in NanoAOD (like `uint16_t`), and made some developments which are I think the first steps towards this. Since https://github.com/cms-sw/cmssw/pull/30273 set the precedent that technical NanoAOD developments are acceptable outside cmssw-nanoAOD, I reopen the PR here almost a year later. Original description in the following.

----------

Hi NanoAOD devs!

I hope this is the right cmssw fork and branch for this PR.

Yesterday I wanted to introduce some new column types in my private nanoAOD productions (int16_t for example, to save a bit of space) and use them in the flat table producers. However, I realized that there are many parts of the NanoAOD code which have to be tweaked if you want to do this, as the way how column types are handled is not completely trivial.

One source of complication is that when you add a column to a flat table with addColumn(), you have to pass the type as a template argument as well as an enum value in the function parameters. After working a bit with the code, I understood that this is redundant, because the check_type function makes sure you always use the right enum value with the right template parameter. Therefore, we could just drop this enum parameter and deduce it from the template argument. In this situation, we would also not need check_type anymore.

The only tricky part are bool columns, which should actually be represented by a uint8_t vector. So far, the logic to take care of this had to be implemented in the plugins that made use of the FlatTable class, but I think I found a way to have this logic directly in the FlatTable class so one can just use addColumn<bool> to create bool columns and they will be internally stored in the uint8_t vector.

What do you think? This simplifies the type handling already quite a bit, and I think it's the good path towards a FlatTable class that will support all basic types that you can also store in TTrees.

I tested this with the local matrix tests so far, can the nano-bot tests still be done here? That would be very cool!

Thanks for considering this and cheers,
Jonas

#### PR validation:

* CMSSW compiles
* local matrix tests pass
* bot tests in cms-nanoAOD pass (see https://github.com/cms-nanoAOD/cmssw/pull/401)

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

No backport  intended.

guitargeek committed:
don't handle column types redundantly anymore

guitargeek committed:
make ColumnType and enum class for type safety

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30436/16539

- This PR adds an extra 56KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/BTagSFProducer.cc  modified in PR(s): #29927, #29937
  - File  PhysicsTools/NanoAOD/plugins/CandMCMatchTableProducer.cc  modified in PR(s): #29933
  - File  PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc  modified in PR(s): #29927, #29933, #29934, #29935, #29926
  - File  PhysicsTools/NanoAOD/plugins/NanoAODDQM.cc  modified in PR(s): #29934
  - File  PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc  modified in PR(s): #29936

cmsbuild commented:
A new Pull Request was created by  @guitargeek (Jonas Rembser) for master.

It involves the following packages:

DataFormats/NanoAOD
PhysicsTools/NanoAOD


@perrotta, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please review it and eventually sign? Thanks.
@gpetruc, @rovere this is something you requested to watch as well.
@silviodonato, @dpiparo you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 028e589ce61a178ba2eaf39d592c6bc2e7444cfb

CMSSW: CMSSW_11_2_X_2020-06-26-1100
SCRAM_ARCH: slc7_amd64_gcc820

The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - #30400 @davidlange6: Add all direct deps to BuildFile for PhysicsTools/PatUtils
 - #30401 @davidlange6: Add all direct deps to BuildFile for PhysicsTools/SelectorUtils
 - #30415 @davidlange6: Add all direct deps to BuildFile for SimDataFormats/ValidationFormats
 - #30417 @davidlange6: Add all direct deps to BuildFile for SimTracker/TrackHistory
 - #30416 @davidlange6: Add all direct deps to BuildFile for SimGeneral/TrackingAnalysis
 - #30198 @davidlange6: Remove CondFormats from DataFormats
 - #30314 @JamminJones: added esConsumes to modules in L1Trigger/L1THGCal
 - #30387 @davidlange6: Add all direct deps to BuildFile for FastSimDataFormats/CTPPSFastSim
 - #30262 @camolezi: Reduce boost dependency in Utilities/XrdAdaptor
 - #30305 @bsunanda: Run4-hgx248 Add ERA corresponding to V12 Geometry
 - #30273 @camolezi: Reduce boost dependency in PhysicsTools/NanoAOD
 - #30351 @slava77: add wf 136.72411 re-miniAOD Run2 UL workflow for Run2016B useful for PPS testing
 - #30288 @silviodonato: Remove GenTtbarCategorizer_cfi.py (moved to fillDescription)
 - #30264 @JamminJones: change EcalTPGScale to use esConsumes
 - #30367 @davidlange6: Add direct deps to BuildFile for SimTracker/TrackerHitAssociation
 - #30392 @davidlange6: Add all direct deps to BuildFile for JetMETCorrections/Modules
 - #30365 @seungjin-yang: Add new modules for GEM offline DQM
 - #30338 @fwyzard: Add tracking-only and pixel tracking-only workflows for Run 3
 - #30388 @davidlange6: Add all direct deps to BuildFile for FastSimDataFormats/External
 - #30361 @guitargeek: edm::Range interface to TrackCandidate recHits

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7448/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7448/git-merge-result

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7448/summary.html

I found follow errors while testing this PR

**Failed tests:  Build  ClangBuild**

* **Build**:

I found compilation error when building: 

<pre>                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-06-26-1100/src/DQMOffline/Muon/src/DiMuonHistograms.cc:2:
/cvmfs/cms-ib.cern.ch/nweek-02634/slc7_amd64_gcc820/external/tbb/2020_U2/include/tbb/task.h:21:139: note: #pragma message: TBB Warning: tbb/task.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.
 #pragma message("TBB Warning: tbb/task.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.")
                                                                                                                                           ^
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-06-26-1100/src/DQMOffline/Muon/src/GEMEfficiencyHarvester.cc: In member function 'void GEMEfficiencyHarvester::doResolution(dqm::legacy::DQMStore::IBooker&, dqm::legacy::DQMStore::IGetter&, std::__cxx11::string)':
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-06-26-1100/src/DQMOffline/Muon/src/GEMEfficiencyHarvester.cc:281:38: error: 'maxEtaPartition_' is not a member of 'GEMeMap'
       res_data[key].reserve(GEMeMap::maxEtaPartition_);
                                      ^~~~~~~~~~~~~~~~
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-06-26-1100/src/DQMOffline/Muon/src/GEMEfficiencyHarvester.cc:307:40: error: 'maxEtaPartition_' is not a member of 'GEMeMap'
         new TH2F(name, title, GEMeMap::maxEtaPartition_, 0.5, GEMeMap::maxEtaPartition_ + 0.5, 2, -0.5, 1.5);
                                        ^~~~~~~~~~~~~~~~
</pre>

* **Clang**:

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>In file included from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-26-1100/src/FWCore/Framework/src/Worker.h:31:
In file included from /cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc820/cms/cmssw-patch/CMSSW_11_2_X_2020-06-26-1100/src/FWCore/Concurrency/interface/WaitingTask.h:25:
/cvmfs/cms-ib.cern.ch/nweek-02634/slc7_amd64_gcc820/external/tbb/2020_U2/include/tbb/task.h:21:9: warning: TBB Warning: tbb/task.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual. [-W#pragma-messages]
#pragma message("TBB Warning: tbb/task.h is deprecated. For details, please see Deprecated Features appendix in the TBB reference manual.")
        ^
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-06-26-1100/src/DQMOffline/Muon/src/GEMEfficiencyHarvester.cc:281:38: error: no member named 'maxEtaPartition_' in 'GEMeMap'
      res_data[key].reserve(GEMeMap::maxEtaPartition_);
                            ~~~~~~~~~^
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_2_X_2020-06-26-1100/src/DQMOffline/Muon/src/GEMEfficiencyHarvester.cc:307:40: error: no member named 'maxEtaPartition_' in 'GEMeMap'
        new TH2F(name, title, GEMeMap::maxEtaPartition_, 0.5, GEMeMap::maxEtaPartition_ + 0.5, 2, -0.5, 1.5);
                              ~~~~~~~~~^
</pre>


The following merge commits were also included on top of IB + this PR after doing git cms-merge-topic: 
 - #30400 @davidlange6: Add all direct deps to BuildFile for PhysicsTools/PatUtils
 - #30401 @davidlange6: Add all direct deps to BuildFile for PhysicsTools/SelectorUtils
 - #30415 @davidlange6: Add all direct deps to BuildFile for SimDataFormats/ValidationFormats
 - #30417 @davidlange6: Add all direct deps to BuildFile for SimTracker/TrackHistory
 - #30416 @davidlange6: Add all direct deps to BuildFile for SimGeneral/TrackingAnalysis
 - #30198 @davidlange6: Remove CondFormats from DataFormats
 - #30314 @JamminJones: added esConsumes to modules in L1Trigger/L1THGCal
 - #30387 @davidlange6: Add all direct deps to BuildFile for FastSimDataFormats/CTPPSFastSim
 - #30262 @camolezi: Reduce boost dependency in Utilities/XrdAdaptor
 - #30305 @bsunanda: Run4-hgx248 Add ERA corresponding to V12 Geometry
 - #30273 @camolezi: Reduce boost dependency in PhysicsTools/NanoAOD
 - #30351 @slava77: add wf 136.72411 re-miniAOD Run2 UL workflow for Run2016B useful for PPS testing
 - #30288 @silviodonato: Remove GenTtbarCategorizer_cfi.py (moved to fillDescription)
 - #30264 @JamminJones: change EcalTPGScale to use esConsumes
 - #30367 @davidlange6: Add direct deps to BuildFile for SimTracker/TrackerHitAssociation
 - #30392 @davidlange6: Add all direct deps to BuildFile for JetMETCorrections/Modules
 - #30365 @seungjin-yang: Add new modules for GEM offline DQM
 - #30338 @fwyzard: Add tracking-only and pixel tracking-only workflows for Run 3
 - #30388 @davidlange6: Add all direct deps to BuildFile for FastSimDataFormats/External
 - #30361 @guitargeek: edm::Range interface to TrackCandidate recHits

You can see more details here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7448/git-log-recent-commits
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7448/git-merge-result

cmsbuild commented:
Comparison not run due to Build errors (RelVals and Igprof tests were also skipped)

silviodonato commented:
please test
#30440 should solve the error

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: 028e589ce61a178ba2eaf39d592c6bc2e7444cfb

CMSSW: CMSSW_11_2_X_2020-06-26-2300
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7455/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
Comparison not run due to Build errors/Fireworks only changes/No short matrix requested (RelVals and Igprof tests were also skipped)

guitargeek committed:
fix clang warning in SimpleFlatTableProducer.h

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30436/16550

- This PR adds an extra 56KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/BTagSFProducer.cc  modified in PR(s): #29927, #29937
  - File  PhysicsTools/NanoAOD/plugins/CandMCMatchTableProducer.cc  modified in PR(s): #29933
  - File  PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc  modified in PR(s): #29927, #29933, #29934, #29935, #29926
  - File  PhysicsTools/NanoAOD/plugins/NanoAODDQM.cc  modified in PR(s): #29934
  - File  PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc  modified in PR(s): #29936

cmsbuild commented:
Pull request #30436 was updated. @perrotta, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

slava77 commented:
please abort
the IB is apparently still broken

cmsbuild commented:
Jenkins tests are aborted.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d4aa20edc28da6e8aa9d1eac1991b59c81b25540
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7459/summary.html
CMSSW: CMSSW_11_2_X_2020-06-27-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7459/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 6 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778915
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778858
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

slava77 commented on DataFormats/NanoAOD/interface/FlatTable.h:
```diff
@@ -104,40 +106,43 @@ namespace nanoaod {
-      vec.insert(vec.end(), values.begin(), values.end());
-      if (type == FloatColumn) {
-        flatTableHelper::MaybeMantissaReduce<T>(mantissaBits).bulk(columnData<T>(columns_.size() - 1));
+      if constexpr (std::is_same<T, bool>()) {
+        columns_.emplace_back(name, docString, ColumnType::Bool, uint8s_.size());
+        uint8s_.insert(uint8s_.end(), values.begin(), values.end());
+      } else if constexpr (std::is_same<T, float>()) {
+        columns_.emplace_back(name, docString, ColumnType::Float, floats_.size());
+        floats_.insert(floats_.end(), values.begin(), values.end());
+        flatTableHelper::MaybeMantissaReduce<float>(mantissaBits).bulk(columnData<float>(columns_.size() - 1));
+      } else {
+        ColumnType type = defaultColumnType<T>();
+        auto &vec = bigVector<T>();
+        columns_.emplace_back(name, docString, type, vec.size());
+        vec.insert(vec.end(), values.begin(), values.end());
```
I'm missing why this if/else specialization is needed.
It looks like the code is already set up to be generic and on L126 we just need
```C++
flatTableHelper::MaybeMantissaReduce<type>(mantissaBits).bulk(columnData<type>(columns_.size() - 1));
```
(or is it "T")

guitargeek replied:
Indeed, you are right. I didn't look at `flatTableHelper::MaybeMantissaReduce` carefully enough.

guitargeek committed:
reorganize more redundant checks in FlatTable

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30436/16608

- This PR adds an extra 64KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/BTagSFProducer.cc  modified in PR(s): #29927, #29937
  - File  PhysicsTools/NanoAOD/plugins/CandMCMatchTableProducer.cc  modified in PR(s): #29933
  - File  PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc  modified in PR(s): #29927, #29933, #29934, #29935, #29926
  - File  PhysicsTools/NanoAOD/plugins/NanoAODDQM.cc  modified in PR(s): #29934
  - File  PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc  modified in PR(s): #29936

cmsbuild commented:
Pull request #30436 was updated. @perrotta, @smuzaffar, @Dr15Jones, @makortel, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

guitargeek commented:
Hi @slava77, thanks for your comment. I had to rebase and force-push because the code was previously based on a broken IB, but all the previous commits up to the last one remained unchanged. I implemented your comment and also found a solution to get rid of the remaining `constexpr if` in `addColumn` by moving the special treatment of the `bool` column to a more fundamental place.

Exceptions related to unsupported type usage have also been replaced with `static_assert`. The way how this is achieved might look a bit strange, but is actually the common solution cited by the cppreference:
https://en.cppreference.com/w/cpp/language/if#Constexpr_If

Local matrix tests pass.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30436/16612

- This PR adds an extra 64KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/BTagSFProducer.cc  modified in PR(s): #29927, #29937
  - File  PhysicsTools/NanoAOD/plugins/CandMCMatchTableProducer.cc  modified in PR(s): #29933
  - File  PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc  modified in PR(s): #29927, #29933, #29934, #29935, #29926
  - File  PhysicsTools/NanoAOD/plugins/NanoAODDQM.cc  modified in PR(s): #29934
  - File  PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc  modified in PR(s): #29936

cmsbuild commented:
Pull request #30436 was updated. @perrotta, @smuzaffar, @Dr15Jones, @makortel, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: a6286283105373cf6ac4bef4fb571c907573f6b3

CMSSW: CMSSW_11_2_X_2020-06-29-2300
SCRAM_ARCH: slc7_amd64_gcc820
You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7516/summary.html

I found follow errors while testing this PR

**Failed tests:  HeaderConsistency**

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7516/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778915
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2778859
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 35 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

guitargeek committed:
throw exceptions in NanoAOD if column type not supported

cmsbuild commented:
The code-checks are being triggered in jenkins.

guitargeek commented:
Hi, I fixed the missing include that caused the header inconsistencies and also added some exceptions that will be thrown if someone would implement more column types but not consider them in all relevant `switch` statements. That's probably better than silently ignoring these columns.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30436/16637

- This PR adds an extra 64KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/BTagSFProducer.cc  modified in PR(s): #29927, #29937
  - File  PhysicsTools/NanoAOD/plugins/CandMCMatchTableProducer.cc  modified in PR(s): #29933
  - File  PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc  modified in PR(s): #29926, #29927, #29933, #29934, #29935
  - File  PhysicsTools/NanoAOD/plugins/NanoAODDQM.cc  modified in PR(s): #29934
  - File  PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc  modified in PR(s): #29936

cmsbuild commented:
Pull request #30436 was updated. @perrotta, @smuzaffar, @Dr15Jones, @makortel, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 32adadf451f63255d29758206527d0a105c6e53a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7541/summary.html
CMSSW: CMSSW_11_2_X_2020-06-30-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7541/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 36
 - DQMHistoTests: Total histograms compared: 2778915
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2778858
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 35 files compared)
 - Checked 152 log files, 16 edm output root files, 36 DQM output files

guitargeek committed:
Avoid use of const_cast with helper function

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30436/16673

- This PR adds an extra 64KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/BTagSFProducer.cc  modified in PR(s): #29927, #29937
  - File  PhysicsTools/NanoAOD/plugins/CandMCMatchTableProducer.cc  modified in PR(s): #29933
  - File  PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc  modified in PR(s): #29926, #29927, #29933, #29934, #29935
  - File  PhysicsTools/NanoAOD/plugins/NanoAODDQM.cc  modified in PR(s): #29934
  - File  PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc  modified in PR(s): #29936

cmsbuild commented:
Pull request #30436 was updated. @perrotta, @smuzaffar, @Dr15Jones, @makortel, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

peruzzim commented:
> This is just a re-opening of [cms-nanoAOD#401](https://github.com/cms-nanoAOD/cmssw/pull/401) from September 2019, but this time to upstream CMSSW instead of cmssw-nanoAOD. Back in the day I was interested in storing columns of so far not supported basic types in NanoAOD (like `uint16_t`), and made some developments which are I think the first steps towards this. Since #30273 set the precedent that technical NanoAOD developments are acceptable outside cmssw-nanoAOD, I reopen the PR here almost a year later. 

@guitargeek I am quite uncomfortable with such a big PR being implemented outside of the cms-nanoAOD repository.
The "precedent" you mention was just touching a couple of files with changes of technical nature, yes.
This PR, instead, is modifying most of the producers with great chance of conflicts with other changes. It is indeed a good moment to do it, because of little concurrent development, but still it should be done through the cms-nanoAOD repository to ensure a full testing for all combinations of datasets and releases, and validate the result.
Also, we probably do want to backport this to avoid future backports to 10_2 and 10_6 to become a nightmare (we will need to maintain those branches for quite a long time still)

guitargeek commented:
Thanks @peruzzim, I completely agree with you. We can discuss offline and once cms-nanoAOD is back in sync, we can continue there. It's nice that we can do the testing in cms-nanoAOD with the review comments by @slava77 already addressed!

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 2b0cd99678c24880f3729da67ea6e009d119e870
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7607/summary.html
CMSSW: CMSSW_11_2_X_2020-07-01-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7607/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2784120
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2784064
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 36 files compared)
 - DQMHistoSizes: changed ( 10224.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 154 log files, 17 edm output root files, 37 DQM output files

slava77 commented on DataFormats/NanoAOD/interface/FlatTable.h:
```diff
@@ -161,25 +159,30 @@ namespace nanoaod {
-    const std::vector<T> &bigVector() const {
-      throw cms::Exception("unsupported type");
+    auto const &bigVector() const {
+      CMS_SA_ALLOW return const_cast<FlatTable *>(this)->bigVector<T>();
```
oddly enough we are still getting 
```const_cast was used, this may result in thread-unsafe code```
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7607/llvm-analysis/report-819fa2.html#EndPath

@makortel @Dr15Jones 
please advise if there is a solution to silence the SA warning

makortel replied:
I suspect the silencing of `const_cast` with `CMS_SA_ALLOW` has not made from an idea to the static analyzer. We could take another look in that, but for time being the best alternative is to add comment explaining why the `const_cast` is acceptable here.

guitargeek replied:
Okay! Should I still keep the `CMS_SA_ALLOW` anyway? It doesn't seem to be harmful, and if it one day also silences `const_cast` it is already there.

I can add a comment in the line before:
```
// justified use of `const_cast` to provide a `const` version of the `bigVector` accessor without code duplication
```

guitargeek replied:
I was just checking if the core guidelines have anything to say about this situation, and actually :+1:

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30436/16734

- This PR adds an extra 16KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/BTagSFProducer.cc  modified in PR(s): #30508, #29927, #29937
  - File  PhysicsTools/NanoAOD/plugins/CandMCMatchTableProducer.cc  modified in PR(s): #29933
  - File  PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc  modified in PR(s): #30508, #29927, #29933, #29934, #29935, #29926
  - File  PhysicsTools/NanoAOD/plugins/NanoAODDQM.cc  modified in PR(s): #30508, #29934
  - File  PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc  modified in PR(s): #29936

cmsbuild commented:
Pull request #30436 was updated. @perrotta, @smuzaffar, @Dr15Jones, @makortel, @cmsbuild, @fgolf, @slava77, @santocch, @peruzzim can you please check and sign again.

slava77 commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: f8d93aa37d9e088c45e9408d830a069a5325cc39
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7658/summary.html
CMSSW: CMSSW_11_2_X_2020-07-02-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7658/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2784120
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2784069
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 154 log files, 17 edm output root files, 37 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/30436	f8d93aa 
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences
- from https://github.com/cms-sw/cmssw/pull/30436#issuecomment-652832213 I expect that this PR is actually not merged directly but instead goes through the cms-nanoAOD repository

santocch commented:
+1

makortel commented on FWCore/Utilities/interface/Range.h:
```diff
@@ -16,6 +18,12 @@ namespace edm {
+    auto const& operator[](std::size_t idx) const { return *(begin_ + idx); }
+
+    auto const& front() const { return *begin_; }
+    auto const& back() const { return *(end_ - 1); }
```
These additions make `edm::Range` to essentially require random-access iterators. They also make `edm::Range` to look more like C++20 [`std::span`](https://en.cppreference.com/w/cpp/container/span) than the C++20 [`std::ranges::range`](https://en.cppreference.com/w/cpp/ranges/range) concept. How about leaving `edm::Range` as it is and create new `edm::Span` with these additions (that would later be replaced with `std::span`)?

guitargeek replied:
Absolutely! I also realized that it now requires random-access iterators to work efficiently, but thought that was okay since we only use `edm::Range` for those (so far). Your idea of making it consistent with `std` is really nice, thank you. I'll do that.

guitargeek committed:
introduce edm::Span

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30436/16859

- This PR adds an extra 20KB to repository

- There are other open Pull requests which might conflict with changes you have proposed:
  - File  PhysicsTools/NanoAOD/plugins/BTagSFProducer.cc  modified in PR(s): #30508, #29927, #29937
  - File  PhysicsTools/NanoAOD/plugins/CandMCMatchTableProducer.cc  modified in PR(s): #29933
  - File  PhysicsTools/NanoAOD/plugins/GenWeightsTableProducer.cc  modified in PR(s): #30508, #29927, #29926, #29933, #29934, #29935
  - File  PhysicsTools/NanoAOD/plugins/NanoAODDQM.cc  modified in PR(s): #30508, #29934
  - File  PhysicsTools/NanoAOD/plugins/TriggerObjectTableProducer.cc  modified in PR(s): #29936

cmsbuild commented:
Pull request #30436 was updated. @perrotta, @smuzaffar, @Dr15Jones, @makortel, @cmsbuild, @gouskos, @fgolf, @slava77, @jpata, @mariadalfonso, @santocch, @peruzzim can you please check and sign again.

makortel commented:
@cmsbuild, please test

cmsbuild commented:
The tests are being triggered in jenkins.

makortel commented on DataFormats/NanoAOD/interface/FlatTable.h:
```diff
@@ -26,7 +27,8 @@ namespace nanoaod {
       }
-      inline void bulk(boost::sub_range<std::vector<float>> data) const {
+      template <typename Span>
+      inline void bulk(Span &&data) const {
```
Why rvalue reference instead of
```suggestion
      inline void bulk(Span const& data) const {
```
?

guitargeek replied:
I once read this blog post about move semantics:
https://eli.thegreenplace.net/2014/perfect-forwarding-and-universal-references-in-c/#id6

Specifically the part "Don't let `T&&` fool you here ...". So it's not a rvalue reference, but makes the function automatically deduce if taking the argument by value or by reference is appropriate. Therefore it appeared to me that it might be a good habit to use `&&` by default for template deduced arguments, especially in this class where `bulk` is used with an rvalue.

Dr15Jones replied:
`T&&` is sometimes referred to as the _universal reference_.

guitargeek replied:
Maybe this isocpp blogpost is a more official reference for these universal references:
https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers

makortel replied:
Oh right, forgot that detail, never mind then.

Dr15Jones replied:
In this particular case I do not see a need for the universal reference as the type is directly interacted with instead of passed to another function via the use of `std::forward`.

guitargeek replied:
That's also true

guitargeek replied:
But it's such a minor change that I think I can also do it later in one go when I rebase this PR for the nanoAOD repository, or if you have more major comments, if this is okay.

Speaking about perfect forwarding, I recently noticed in the PR where I used FWCore/SOA that the SOA Table might make several unnecessary copies because even though `std::forward` is used, this is not matched by the corresponding universal reference for the argument, for example here (and also in other functions for the implementation details):
https://github.com/cms-sw/cmssw/blob/master/FWCore/SOA/interface/Table.h#L145

I didn't open an issue about this because I didn't think it was that important, but I it's just something to keep in mind for the next time anyone will work with FWCore/SOA.

makortel replied:
> But it's such a minor change that I think I can also do it later in one go when I rebase this PR for the nanoAOD repository, or if you have more major comments, if this is okay.

Perfectly fine for me, it is minor indeed.

> I didn't open an issue about this because I didn't think it was that important, but I it's just something to keep in mind for the next time anyone will work with FWCore/SOA.

The best way to remind us (`core`) is to open an issue :)

guitargeek replied:
Ok thanks!

cmsbuild commented:
+1
Tested at: ea4170210f0738ac0630d4f1856988c3f5324a62
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7789/summary.html
CMSSW: CMSSW_11_2_X_2020-07-08-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/7789/summary.html

The workflows 136.7611, 136.731, 4.53, 4.22 have different files in step1_dasquery.log than the ones found in the baseline. You may want to check and retrigger the tests if necessary. You can check it in the "files" directory in the results of the comparisons

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 37
 - DQMHistoTests: Total histograms compared: 2787364
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2787313
 - DQMHistoTests: Total skipped: 50
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 36 files compared)
 - Checked 154 log files, 17 edm output root files, 37 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/30436	ea41702 
- changes since the last signoff are technical for reco (requested by core, to move from what was proposed to be edm::Range to become edm::Span)
- tests pass and comparisons show no differences

makortel commented:
+1

santocch commented:
+1

mariadalfonso commented:
@guitargeek 
I'm in favour of having the technical PR merged-reviews in the cms-sw (rather than in cms-nano).

1. Can you update the,  cms-nanoAOD#401 so that I test the nano-bot
2. do you need to feature in the 10_6 as well (release that likely will be used for the Run2-nano-UL ?)

guitargeek commented:
Hi @mariadalfonso, thank you very much for making a decision, it sound very good to me!

I will update the cms-nanoAOD PR and implement the latest comment from @makortel about the universal reference in the cms-sw PR so it can be merged.

About the backports: @peruzzim mentioned in his last comment (https://github.com/cms-sw/cmssw/pull/30436#issuecomment-652832213) that both the 10_2_X and 10_6_X cycles still need to be kept in sync with the recent NanoAOD developments. That would mean that this PR has to be backported to both 10_2_X and 10_6_X, no?

guitargeek committed:
change unneeded universal reference to const&

cmsbuild commented:
The code-checks are being triggered in jenkins.

guitargeek commented:
Once again a rebase because the IB I based my developments on expired. The only actual difference is the last added commit.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-30436/17191

- This PR adds an extra 68KB to repository

cmsbuild commented:
Pull request #30436 was updated. @perrotta, @smuzaffar, @Dr15Jones, @makortel, @cmsbuild, @gouskos, @fgolf, @slava77, @jpata, @mariadalfonso, @santocch, @peruzzim can you please check and sign again.

guitargeek commented:
PR https://github.com/cms-nanoAOD/cmssw/pull/401 is now in sync with this PR.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

guitargeek commented:
I just tested how complicated a backport to 10_6_X would be, and it would already be a nightmare due to the clang-format PR https://github.com/cms-sw/cmssw/pull/27353. There was already some manual editing done in previous backports to circumvent the absence of clang-format changes in older releases, so at this point the clang-format commit can't be trivially backported anymore.

I think this is not a good situation, because it stalls NanoAOD developments. People are on one hand reluctant to do anything for master because the backports are so terrible, and on the other hand it's also a nightmare to "forward-port" NanoAOD developments that you did in e.g. 10_2_X to improve NanoAOD for your analysis. There are for example reported problems in the hypernews like [here]<https://hypernews.cern.ch/HyperNews/CMS/get/generators/4710.html> that everyone (also me) seems to solve for herself in 10_2_X.

In conclusion: I can't do any backports of this PR without manual changes, which would only make the situation worse. Therefore, I advise against merging my own PR before the fundamental NanoAOD backporting problem is solved. Backporting the clang-format commit might be a good start.

slava77 commented:
> In conclusion: I can't do any backports of this PR without manual changes, which would only make the situation worse. Therefore, I advise against merging my own PR before the fundamental NanoAOD backporting problem is solved. Backporting the clang-format commit might be a good start.

What are the benefits of backporting (even to 106X)?
Can Nano move to a more recent production release, e.g. 11_1_X?

cmsbuild commented:
+1
Tested at: 00b26e72b56c4f5661d8bb074b59edb209b3a232
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/8175/summary.html
CMSSW: CMSSW_11_2_X_2020-07-21-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce380e/8175/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 34
 - DQMHistoTests: Total histograms compared: 2525996
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2525948
 - DQMHistoTests: Total skipped: 47
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 33 files compared)
 - Checked 144 log files, 17 edm output root files, 34 DQM output files

slava77 commented:
+1

for https://github.com/cms-sw/cmssw/pull/30436	00b26e7 
- rebased since the last signoff for https://github.com/cms-sw/cmssw/pull/30436	ea41702 
with an addition of somewhat trivial update in 00b26e72b56c4f5661d8bb074b59edb209b3a232
- jenkins tests pass and comparisons show no differences

makortel commented:
+1

mariadalfonso commented:
+1

mariadalfonso commented:
> > In conclusion: I can't do any backports of this PR without manual changes, which would only make the situation worse. Therefore, I advise against merging my own PR before the fundamental NanoAOD backporting problem is solved. Backporting the clang-format commit might be a good start.
> 
> What are the benefits of backporting (even to 106X)?
> Can Nano move to a more recent production release, e.g. 11_1_X?

For the analysis or dedicated production, I think the 11_1_X can be used indeed
But when we will do the UL-reminiAOD I immagine the 106X and at the same time some version of nano will be produced.

We are planning a XPOG meeting this Tuesday with a general discussion
https://indico.cern.ch/event/939863

guitargeek commented:
Thank you very much!

slava77 commented:
> > > In conclusion: I can't do any backports of this PR without manual changes, which would only make the situation worse. Therefore, I advise against merging my own PR before the fundamental NanoAOD backporting problem is solved. Backporting the clang-format commit might be a good start.
> > 
> > 
> > What are the benefits of backporting (even to 106X)?
> > Can Nano move to a more recent production release, e.g. 11_1_X?
> 
> For the analysis or dedicated production, I think the 11_1_X can be used indeed
> But when we will do the UL-reminiAOD I immagine the 106X and at the same time some version of nano will be produced.
> 
> We are planning a XPOG meeting this Tuesday with a general discussion
> https://indico.cern.ch/event/939863

I wish we didn't have to backport so much stuff to 10_6_X for reminiAOD either, since we actually support remini in the master a more backport-friendly recent production release could be (have been) used.
At least we should minimize needs to backport to 10_2_X as well.
Let's see what comes out from the meeting discussion.

santocch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @silviodonato, @dpiparo (and backports should be raised in the release meeting by the corresponding L2)

silviodonato commented:
@mariadalfonso @gouskos @peruzzim Can I merge this PR? Shall I wait for cms-nanoAOD#401 ?

guitargeek commented:
Hi @silviodonato , I think Mariarosaria was pretty clear in https://github.com/cms-sw/cmssw/pull/30436#issuecomment-661719525 and especially https://github.com/cms-nanoAOD/cmssw/pull/401#issuecomment-662406510 that this PR to cms-sw is the one to merge.

silviodonato commented:
+1
